File: web-backend/src/main/kotlin/velocorner/backend/util/ActivityOps.kt
package velocorner.backend.util

import velocorner.backend.model.*
import java.time.LocalDate

object ActivityOps {

    fun toSumYtdSeries(
        activities: Sequence<Activity>,
        now: LocalDate,
        action: String,
        unit: Units
    ): List<DailySeries> {
        val dailyProgress = DailyProgress.from(activities)
        val yearlyProgress = YearlyProgress.from(dailyProgress)
        val ytdSum = YearlyProgress.sumYtd(yearlyProgress, now)

        return when (action.lowercase()) {
            "distance" -> Highcharts.toDistanceSeries(ytdSum, unit)
            "elevation" -> Highcharts.toElevationSeries(ytdSum, unit)
            "time" -> Highcharts.toTimeSeries(ytdSum, unit)
            else -> error("not supported action: $action")
        }
    }

    fun toYearlySeries(
        activities: Sequence<Activity>,
        action: String,
        unit: Units
    ): List<DailySeries> {
        val dailyProgress = DailyProgress.from(activities)
        val yearlyProgress = YearlyProgress.from(dailyProgress)

        return when (action.lowercase()) {
            "heatmap" -> Highcharts.toDistanceSeries(YearlyProgress.zeroOnMissingDate(yearlyProgress), unit)
            "distance" -> Highcharts.toDistanceSeries(YearlyProgress.aggregate(yearlyProgress), unit)
            "elevation" -> Highcharts.toElevationSeries(YearlyProgress.aggregate(yearlyProgress), unit)
            "time" -> Highcharts.toTimeSeries(YearlyProgress.aggregate(yearlyProgress), unit)
            else -> error("not supported action: $action")
        }
    }
}

File: web-app/app/controllers/util/ActivityOps.scala
package controllers.util

import model.highcharts
import org.joda.time.LocalDate
import velocorner.api.Units
import velocorner.api.chart.DailySeries
import velocorner.api.strava.Activity
import velocorner.model.{DailyProgress, YearlyProgress}

trait ActivityOps {

  def toSumYtdSeries(activities: Iterable[Activity], now: LocalDate, action: String, unit: Units.Entry): List[DailySeries] = {
    val dailyProgress = DailyProgress.from(activities)
    val yearlyProgress = YearlyProgress.from(dailyProgress)
    val ytdSum = YearlyProgress.sumYtd(yearlyProgress, now)
    action.toLowerCase match {
      case "distance"  => highcharts.toDistanceSeries(ytdSum, unit)
      case "elevation" => highcharts.toElevationSeries(ytdSum, unit)
      case "time"      => highcharts.toTimeSeries(ytdSum, unit)
      case other       => sys.error(s"not supported action: $other")
    }
  }

  def toYearlySeries(activities: Iterable[Activity], action: String, unit: Units.Entry): List[DailySeries] = {
    val dailyProgress = DailyProgress.from(activities)
    val yearlyProgress = YearlyProgress.from(dailyProgress)
    action.toLowerCase match {
      case "heatmap"   => highcharts.toDistanceSeries(YearlyProgress.zeroOnMissingDate(yearlyProgress), unit)
      case "distance"  => highcharts.toDistanceSeries(YearlyProgress.aggregate(yearlyProgress), unit)
      case "elevation" => highcharts.toElevationSeries(YearlyProgress.aggregate(yearlyProgress), unit)
      case "time"      => highcharts.toTimeSeries(YearlyProgress.aggregate(yearlyProgress), unit)
      case other       => sys.error(s"not supported action: $other")
    }
  }
}
